assembler/asm2byte/asm2byte.cpp

```diff
@@ -0,0 +1,260 @@
+#include "asm2byte.h"
+
+#include "common/str_to_opcode.h"
+#include "common/utils/string_operations.h"
+
+#include <fstream>
+#include <iostream>
+#include <algorithm>
+#include <iostream>
+
+namespace evm::asm2byte {
+
+bool AsmToByte::ParseAsm()
+{
+    PrepareLinesFromBuffer();
+
+    if (!CreateInstructionsFromLines()) {
+        std::cerr << "Error in CreateInstructionsFromLines" << std::endl;
+        return false;
+    }
+
+    return true;
+}
+
+bool AsmToByte::ParseAsmFile(const char *filename)
+{
+    if (!OpenBytecodeFile(filename)) {
+        return false;
+    }
+
+    return ParseAsm();
+}
+
+bool AsmToByte::ParseAsmString(const std::string& asm_string)
+{
+    file_buffer_ = asm_string;
+    return ParseAsm();
+}
+
+bool AsmToByte::DumpInstructionsToBytes()
+{
+    for (auto it : instructions_) {
+        it->InstrToBytes(&bytecode_);
+    }
+    return true;
+}
+
+bool AsmToByte::DumpBytesInBytecode(const char *filename)
+{
+    std::ofstream outfile(filename, std::ios::out | std::ios::binary);
+    outfile.write(reinterpret_cast<char *>(bytecode_.data()), bytecode_.size());
+    return true;
+}
+
+bool AsmToByte::OpenBytecodeFile(const char *filename)
+{
+    if (filename == nullptr) {
+        std::cerr << "Filename is nullptr" << std::endl;
+        return false;
+    }
+
+    std::ifstream file(filename, std::ios::binary | std::ios::ate);
+    if (!file) {
+        std::cerr << "Error when opening file " << filename << std::endl;
+        return false;
+    }
+
+    std::streamsize filesize = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    file_buffer_.resize(filesize);
+
+    if (!file.read(file_buffer_.data(), filesize)) {
+        std::cerr << "Can not read file in buffer" << std::endl;
+        return false;
+    }
+
+    file.close();
+
+    return true;
+}
+
+void AsmToByte::PrepareLinesFromBuffer()
+{
+    bool is_string_endline_terminated = (file_buffer_.back() == '\n');
+
+    std::size_t line_start_idx = 0;
+
+    size_t i = 0;
+    for (; std::isspace(file_buffer_[i]) != 0; ++i) {
+        ++line_start_idx;
+    }
+
+    for (; i < file_buffer_.size(); ++i) {
+        if (std::isspace(file_buffer_[i]) != 0 || file_buffer_[i] == ',') {
```

Comment:
`else (file_buffer_[i] == ';')` - what will be the result in this case?

Reply:
Is `std::isspace(file_buffer_[i]) != 0` necessary in this place?

Reply:
Added TODO